common/src/main/java/io/netty/util/DefaultAttributeMap.java

```diff
@@ -17,186 +17,240 @@
-            head = new DefaultAttribute();
-            DefaultAttribute<T> attr = new DefaultAttribute<T>(head, key);
-            head.next = attr;
-            attr.prev = head;
-            if (attributes.compareAndSet(i, null, head)) {
-                // we were able to add it so return the attr right away
-                return attr;
-            } else {
-                head = attributes.get(i);
+        return -(low + 1);
+    }
+
+    private static void orderedCopyOnInsert(DefaultAttribute[] sortedSrc, int srcLength, DefaultAttribute[] copy,
+                                            DefaultAttribute toInsert) {
+        // let's walk backward, because as a rule of thumb, toInsert.key.id() tends to be higher for new keys
+        final int id = toInsert.key.id();
+        int i;
+        for (i = srcLength - 1; i >= 0; i--) {
+            DefaultAttribute attribute = sortedSrc[i];
+            assert attribute.key.id() != id;
+            if (attribute.key.id() < id) {
+                break;
             }
+            copy[i + 1] = sortedSrc[i];
+        }
+        copy[i + 1] = toInsert;
+        final int toCopy = i + 1;
+        if (toCopy > 0) {
+            System.arraycopy(sortedSrc, 0, copy, 0, toCopy);
         }
+    }
 
-        synchronized (head) {
-            DefaultAttribute<?> curr = head;
-            for (;;) {
-                DefaultAttribute<?> next = curr.next;
-                if (next == null) {
-                    DefaultAttribute<T> attr = new DefaultAttribute<T>(head, key);
-                    curr.next = attr;
-                    attr.prev = curr;
-                    return attr;
-                }
+    private volatile DefaultAttribute[] attributes = EMPTY_ATTRIBUTES;
 
-                if (next.key == key && !next.removed) {
-                    return (Attribute<T>) next;
+    @SuppressWarnings("unchecked")
+    @Override
+    public <T> Attribute<T> attr(AttributeKey<T> key) {
+        ObjectUtil.checkNotNull(key, "key");
+        DefaultAttribute newAttribute = null;
+        for (;;) {
+            final DefaultAttribute[] attributes = this.attributes;
+            final int count = attributes.length;
+            final int index = searchAttributeByKey(attributes, 0, count, key);
+            final DefaultAttribute[] newAttributes;
+            if (index >= 0) {
+                final DefaultAttribute attribute = attributes[index];
+                assert attribute.key() == key;
+                if (!attribute.removed()) {
+                    return attribute;
+                }
+                // let's try replace the removed attribute with a new one
+                if (newAttribute == null) {
+                    newAttribute = new DefaultAttribute<T>(this, key);
+                }
+                newAttributes = Arrays.copyOf(attributes, count);
+                newAttributes[index] = newAttribute;
+            } else {
```

Comment:
This doesn't seem to take advantage of the `index` which should be the insertion point.    I would suspect after a few attrs, avoiding the excess copies would become expensive.    It feels like the index should be passed into `orderedCopyOnInsert `